### PR TITLE
Use STM32 Timer channel signal in methods

### PR DESCRIPTION
**Pull request by strongly-typed**
Adding on @chris-durand example in #908

The `advanced_extended` property still looks a bit rough to me. That information should come from modm-devices ... I had to disable some device families to get all tests run. I would leave it as it is and add better props later.

### CODE REVIEW

**Comment by strongly-typed** [comment]
@salkinium I would love to set the `ci:hal` label so the CI tests all devices.

File: src/modm/platform/timer/stm32/module.lb [hunk 1]
```diff
@@ -110,8 +110,8 @@ def prepare(module, options):
     props["target"] = target = device.identifier
     props["types"] = set()
     # extended advanced control timer with 6 channels and trigger out 2
-    props["advanced_extended"] = target.family not in ["f0", "f1", "f2", "f4", "l0", "l1"] or \
-            target.family == "f3" and target.name not in ["73", "78"]
+    props["advanced_extended"] = \
+            (target.family == "f3" and target.name not in ["03", "73", "78"])
```
Comment (chris-durand) [inline src/modm/platform/timer/stm32/module.lb]: Nope. This breaks my production code :wink:. The advanced control timer implementation with 6 channels is present in all F7, G0, G4, H7, L4, L5, U5, WB and WL devices as well as all F3 except 373 and 378.

If you remove the `advanced extended` property user code using timer channels 5 and 6 will still compile and access wrong places in the peripheral memory when trying to configure those channels. That would be very bad there, for instance:

```
	static inline void
	setCompareValue(uint32_t channel, Value value)
	{
%% if advanced_extended
		if(channel <= 4) {
			*(&TIM{{ id }}->CCR1 + (channel - 1)) = value;
		} else if(channel == 5) {
			TIM{{ id }}->CCR5 = value;
		} else if(channel == 6) {
			TIM{{ id }}->CCR6 = value;
		}
%% else
		*(&TIM{{ id }}->CCR1 + (channel - 1)) = value;
%% endif
	}
```

**Review by chris-durand — CHANGES_REQUESTED** [review]

File: src/modm/platform/timer/stm32/module.lb (hunk 1 above)
Comment (chris-durand) [reply src/modm/platform/timer/stm32/module.lb]: From looking at the ST headers the correct condition should be:
```suggestion
    props["advanced_extended"] = target.family not in ["f0", "f1", "f2", "f3", "f4", "l0", "l1"] and \
            (target.family != "f3" or target.name not in ["73", "78"])
```

If there is a compile error for F303, then all what we call `advanced_extended` is not actually identical.

**Comment by salkinium** [comment]
> I would love to set the ci:hal label so the CI tests all devices.

I've given you triage rights for this repo, so you should be able to set issues/pr labels and milestones yourself.

File: src/modm/platform/timer/stm32/module.lb (hunk 1 above)
Comment (strongly-typed) [reply src/modm/platform/timer/stm32/module.lb]: ... but GPIO `base.hpp` is missing `Ch5` and `Ch6` in many cases:

```
/// @cond
	enum class
	Signal
	{
[...]
		Ch1,
		Ch1n,
		Ch2,
		Ch2n,
		Ch3,
		Ch3n,
		Ch4,
		Ch4n,
		Cha1,
		Cha2,
		Chb1,
[...]
```

e.g. for `examples/nucleo_g474re/servo_pwm`
```
Compiling C++·· release/modm/src/modm/platform/timer/timer_4.o
In file included from modm/src/modm/platform/timer/timer_4.hpp:23,
                 from modm/src/modm/platform/timer/timer_4.cpp:17:
modm/src/modm/platform/timer/general_purpose_base.hpp: In static member function 'static consteval int modm::platform::GeneralPurposeTimer::signalToChannel()':
modm/src/modm/platform/timer/general_purpose_base.hpp:272:56: error: 'Ch5' is not a member of 'modm::platform::Gpio::Signal'; did you mean 'Ch1'?
  272 |   } else if constexpr (Signal::Signal == Gpio::Signal::Ch5) {
```

Comment (chris-durand) [reply src/modm/platform/timer/stm32/module.lb]: The CI compile error is caused by the absence of `Gpio::Signal::Ch5` and `Gpio::Signal::Ch6`. Channels 5 and 6 can't be connected to IO pins on any STM32, so there are no GPIO signals for it. That was my mistake. Just remove the following lines in `signalToChannel()` and everything should work:

```
%% if advanced_extended
		} else if constexpr (Signal::Signal == Gpio::Signal::Ch5) {
			return 5;
		} else if constexpr (Signal::Signal == Gpio::Signal::Ch6) {
			return 6;
%% endif
```

Comment (strongly-typed) [reply src/modm/platform/timer/stm32/module.lb]: > Channels 5 and 6 can't be connected to IO pins on any STM32, so there are no GPIO signals for it.

I was just studying the datasheets and came to the same conclusion. No worries, just fixed it.

**Comment by chris-durand** [comment]
I am sorry, I messed it up again :facepalm:. The correct condition for "advanced_extended" is:
```
    props["advanced_extended"] = target.family not in ["f0", "f1", "f2", "f3", "f4", "l0", "l1"] or \
            (target.family == "f3" and target.name not in ["73", "78"])
```
Otherwise "advanced_extended" will never be true on F3.

**Review by chris-durand — APPROVED** [review]
Thank you!

**Comment by strongly-typed** [comment]
The 14-pin versions of [STM32L011D3P6](https://github.com/modm-io/modm/runs/8287421849?check_suite_focus=true#step:5:39) fail in the CI as they only have 3 channels.

What is the best way to distinguish these parts?

Something like?
```
    props["advanced_extended"] = ((target.family not in ["f0", "f1", "f2", "f3", "f4", "l0", "l1"]) and (target.package != 'd')) or \
            ((target.family == "f3") and (target.name not in ["73", "78"]))
```

**Comment by chris-durand** [comment]
> The 14-pin versions of STM32L011D3P6 fail in the CI as they only have 3 channels.
>
> What is the best way to distinguish these parts?

`advanced_extended` is not the issue here, the condition is false already for `L0`. The devices have only one general purpose and one basic timer.

A hacky but easy solution would be guarding the `Ch4` if-branch with a condition to exclude them:

```
%% if target.family != "l0" or target.package != "d"
		} else if constexpr (Signal::Signal == Gpio::Signal::Ch4) {
			return 4;
%% endif
```

Or you determine from the gpio data which signals exist. You would need to add a query to the gpio module to get the signals from the timer driver and only generate code in `signalToChannel()` for the existing channels.

I am fine with the easy one. If ST released more tiny devices with missing timer channels someone would have to touch that condition again. But that should not be a big deal.

**Comment by strongly-typed** [comment]
> I am fine with the easy one. If ST released more tiny devices with missing timer channels someone would have to touch that condition again. But that should not be a big deal.

Let's go with the easy solution first and see where it takes us.

I would like to ask if you could share any example code that uses `Ch5` and `Ch6` in the CI? Currently, this is all unused and further changes are not checked by the CI.

**Comment by chris-durand** [comment]
> Let's go with the easy solution first and see where it takes us.

After testing it, I know we won't come far :laughing:. It's `target.pin != "d"`, the package part in `STM32L011D3P6` is the `p`.

**Comment by chris-durand** [comment]
> I would like to ask if you could share any example code that uses `Ch5` and `Ch6` in the CI? Currently, this is all unused and further changes are not checked by the CI.

I am actually planning to add an example once I have implemented the ADC external trigger function in the F3/G4/H7/... ADC driver. It currently would not add much value to the CI. The code is not much different to using channels 1-4. You only pass a different number to `configureOutputChannel()` and that is a runtime parameter not checked during compilation:
```
Motor::PwmTimer::configureOutputChannel(5, PwmMode, CurrentSamplingOffset);
```
Then it configures TRGO2 to output the channel 5 compare signal (by calling `setMode()`). That signal triggers the ADCs.

Right now I am writing ADC registers manually, but it should not be too hard to add that API to the modm driver for all devices (if you can parse the list of triggers from the reference manuals, but I have heard @salkinium has a solution to that problem :smiley:).

**Comment by chris-durand** [comment]
I will push the fix with `target.pin != "d"` instead of `target.package != "d"` to your branch and restart the compile all job.

**Comment by strongly-typed** [comment]
The CI ran fine. So if there aren't any objections: LGTM. Thank you all for your good support and great collaboration.

**Review by salkinium — APPROVED** [review]
Will merge tonight, currently traveling.

**Pull request merged**